cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -388,101 +390,141 @@ func (c ControllerContext) IsControllerEnabled(name string) bool {
 // that requests no additional features from the controller manager.
 // Any error returned will cause the controller process to `Fatal`
 // The bool indicates whether the controller was enabled.
-type InitFunc func(ctx context.Context, controllerCtx ControllerContext) (controller controller.Interface, enabled bool, err error)
+type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+
+type ControllerDescriptor struct {
+	name                      string
+	initFunc                  InitFunc
+	requiredFeatureGates      []featuregate.Feature
+	isDisabledByDefault       bool
+	isCloudProviderController bool
+}
+
+func (r *ControllerDescriptor) Name() string {
+	return r.name
+}
+
+func (r *ControllerDescriptor) GetInitFunc() InitFunc {
+	return r.initFunc
+}
+
+func (r *ControllerDescriptor) GetRequiredFeatureGates() []featuregate.Feature {
+	return append([]featuregate.Feature(nil), r.requiredFeatureGates...)
+}
 
-// ControllerInitializersFunc is used to create a collection of initializers
+func (r *ControllerDescriptor) IsDisabledByDefault() bool {
+	return r.isDisabledByDefault
+}
+
+func (r *ControllerDescriptor) IsCloudProviderController() bool {
+	return r.isCloudProviderController
+}
+
+// ControllerDescriptorsFunc is used to create a collection of controller descriptors
 // given the loopMode.
-type ControllerInitializersFunc func(loopMode ControllerLoopMode) (initializers map[string]InitFunc)
+type ControllerDescriptorsFunc func() (initializers map[string]*ControllerDescriptor)
```

Comment:
"initializers" seems like a weird name for this - why name the return at all?

Reply:
True. Although I have noticed that after the changes in the https://github.com/kubernetes/kubernetes/pull/120371/commits/a85779b4dfe4c71a2def2f2385fb4d8f896522ad commit, we do not need the function type at all. So I have removed it.